lib/snapshot-eviction/src/evictor.rs

```diff
@@ -0,0 +1,276 @@
+    }
+
+    /// Query for eviction candidates
+    #[instrument(name = "snapshot_evictor.find_candidates", skip(self))]
+    async fn find_candidates(
+        &self,
+    ) -> SnapshotEvictionResult<Vec<(WorkspaceSnapshotAddress, SystemTime)>> {
+        let client = self.si_db_pool.get().await?;
+
+        let rows = client
+            .query(
+                "SELECT s.snapshot_id, s.last_used_at
+                 FROM snapshot_last_used s
+                 LEFT JOIN change_set_pointers cs ON cs.workspace_snapshot_address = s.snapshot_id
+                 WHERE s.last_used_at < (NOW() - $1 * INTERVAL '1 second')
+                   AND cs.id IS NULL
+                 ORDER BY s.last_used_at ASC
+                 LIMIT $2
+                 FOR UPDATE OF s SKIP LOCKED",
+                &[
+                    &(self.config.grace_period_seconds as f64),
+                    &(self.config.batch_size as i64),
+                ],
+            )
+            .await?;
+
+        let mut candidates = Vec::with_capacity(rows.len());
+        for row in rows {
+            let address: WorkspaceSnapshotAddress = row.try_get("snapshot_id")?;
+            let last_used_at: SystemTime = row.try_get("last_used_at")?;
+            candidates.push((address, last_used_at));
+        }
+
+        // Record metric - histogram to track distribution of workload characteristics
+        histogram!(snapshot_eviction_candidates_found = candidates.len() as f64);
+
+        Ok(candidates)
+    }
+
+    /// Evict a single snapshot with metrics tracking
+    async fn evict_snapshot(
+        &self,
+        address: &WorkspaceSnapshotAddress,
+        last_used_at: SystemTime,
+    ) -> SnapshotEvictionResult<()> {
+        // Calculate candidate queue latency
+        let grace_period = Duration::from_secs(self.config.grace_period_seconds as u64);
+        let eligible_at = last_used_at + grace_period;
+        let now = SystemTime::now();
+
+        if let Ok(queue_time) = now.duration_since(eligible_at) {
+            histogram!(snapshot_candidate_queue_latency_seconds = queue_time.as_secs_f64());
+        }
+
+        let start = Instant::now();
+
+        match self.evict_snapshot_inner(address).await {
+            Ok(()) => {
+                // Record success duration (existing metric, already in evict_snapshot_inner)
+                Ok(())
+            }
+            Err(e) => {
+                // Extract error type for labeling
+                let error_type = match &e {
+                    crate::SnapshotEvictionError::Database(_) => "database",
+                    crate::SnapshotEvictionError::PgPool(_) => "pg_pool",
+                    crate::SnapshotEvictionError::LayerCache(_) => "layer_cache",
+                    crate::SnapshotEvictionError::Nats(_) => "nats",
+                };
+
+                // Record failure counter with error type
+                monotonic!(snapshot_eviction_failures = 1, error_type = error_type);
```

Comment:
That's a really nice looking pattern. After we do this 3,4,5 times I bet there's some kind of pattern extraction (even if it's a macro) to make this almost a one-liner